From 814af2e1de14a017b0cb78a14d804294ad66de34 Mon Sep 17 00:00:00 2001
From: Maximiliano Sandoval R <msandova@gnome.org>
Date: Fri, 6 Oct 2023 17:30:41 +0200
Subject: [PATCH 2/2] notification: Use g_signal_connect_object

This ensures the notification lives as long as the signal. This was
added in 2.62 so we add that to meson.build.

Fixes https://gitlab.gnome.org/GNOME/libnotify/-/issues/34
---
 libnotify/notification.c | 21 +++++++--------------
 meson.build              |  2 +-
 2 files changed, 8 insertions(+), 15 deletions(-)

diff --git a/libnotify/notification.c b/libnotify/notification.c
index 3fd261a..0240f62 100644
--- a/libnotify/notification.c
+++ b/libnotify/notification.c
@@ -405,8 +405,8 @@ notify_notification_dispose (GObject *object)
         }
 
         proxy = _notify_get_proxy (NULL);
-        if (proxy != NULL && priv->proxy_signal_handler != 0) {
-                g_signal_handler_disconnect (proxy, priv->proxy_signal_handler);
+        if (proxy != NULL) {
+                g_clear_signal_handler (&priv->proxy_signal_handler, proxy);
         }
 
         g_clear_object (&priv->icon_pixbuf);
@@ -684,20 +684,12 @@ activate_action (NotifyNotification *notification,
                 return FALSE;
         }
 
-        /* Some clients have assumed it is safe to unref the
-         * Notification at the end of their NotifyActionCallback
-         * so we add a temporary ref until we're done with it.
-         */
-        g_object_ref (notification);
-
         notification->priv->activating = TRUE;
         pair->cb (notification, (char *) action, pair->user_data);
         notification->priv->activating = FALSE;
         g_free (notification->priv->activation_token);
         notification->priv->activation_token = NULL;
 
-        g_object_unref (notification);
-
         return TRUE;
 }
 
@@ -1107,10 +1099,11 @@ notify_notification_show (NotifyNotification *notification,
         }
 
         if (priv->proxy_signal_handler == 0) {
-                priv->proxy_signal_handler = g_signal_connect (proxy,
-                                                               "g-signal",
-                                                               G_CALLBACK (proxy_g_signal_cb),
-                                                               notification);
+                priv->proxy_signal_handler = g_signal_connect_object (proxy,
+                                                                      "g-signal",
+                                                                      G_CALLBACK (proxy_g_signal_cb),
+                                                                      notification,
+                                                                      G_CONNECT_DEFAULT);
         }
 
         if (_notify_uses_portal_notifications ()) {
diff --git a/meson.build b/meson.build
index 169d596..169f66c 100644
--- a/meson.build
+++ b/meson.build
@@ -39,7 +39,7 @@ man1dir = join_paths(prefix, get_option('mandir'), 'man1')
 libnotify_deps = []
 extra_deps = []
 
-glib_req_version = '>= 2.38.0'
+glib_req_version = '>= 2.62.0'
 
 gdk_pixbuf_dep = dependency('gdk-pixbuf-2.0')
 glib_dep = dependency('glib-2.0', version: glib_req_version)
-- 
2.41.0

